src/output.cpp

```diff
@@ -1412,72 +1412,51 @@ std::string from_sentence_case (const std::string &kingston)
-            // Buffer size was sufficient, string has been printed, all is well
+    std::string buffer(result, '\0');
+    _vsprintf_p(&buffer[0], result + 1, format, args); //+1 for string's null
+
+    return buffer;
+}
+#else
+std::string vstring_format(char const *const format, va_list args)
+{
+    errno = 0; // Clear errno before trying
+    std::vector<char> buffer(1024, '\0');
+
+    while (size_t const buf_size = buffer.size()) { // aka while (true)
+        va_list args_copy;
+        va_copy(args_copy, args);
+        int const result = vsnprintf(&buffer[0], buf_size, format, args_copy);
+        va_end(args_copy);
+
+        // All is well.
+        if (result >= 0 && static_cast<size_t>(result) < buf_size) {
             break;
-        } else if( returned_length > 0 ) {
-            // For some reason (is this a mingw build with mingws own vsnprintf?)
-            // vsnprintf seems to be POSIX compatible and returns the required
-            // size of the buffer instead of -1
-            // Note that buffer_size is always > 0 and therefor the case returned_length==0
-            // is handled above.
-            buffer_size = returned_length + 1;
-        } else {
-            buffer_size *= 2;
         }
+
+        // Standards conformant versions return -1 on error only.
+        // Some non-standard versions return -1 to indicate a bigger buffer is needed.
+        if (result < 0 && errno) {
+            return std::string("Bad format string for printf.");
+        }
+
+        // Looks like we need to grow... bigger, definitely bigger.
+        buffer.resize(buf_size * 2);
     }
-#else
-    va_copy(cur_argptr, argptr);
-    const int required = vsnprintf(&buffer[0], buffer_size, pattern, cur_argptr);
-    va_end(cur_argptr);
-    if (required < 0) {
-        return std::string("invalid input to string_format function!");
-    } else if (required >= buffer_size) {
-        // Did not fit the buffer, retry with better buffer size.
-        buffer_size = required + 1;
-        buffer.resize(buffer_size, '\0');
-        // Try again one time, this should be save as we know the required
-        // buffer size and have allocated that much.
-        va_copy(cur_argptr, argptr);
-        vsnprintf(&buffer[0], buffer_size, pattern, cur_argptr);
-        va_end(cur_argptr);
-        // ignore the result of vsnprintf, because it returns different
-        // things on windows, see above.
-        returned_length = required;
-    } else {
-        returned_length = required;
-    }
-#endif
-    //drop contents behind \003, this trick is there to skip certain arguments
-    std::vector<char>::iterator a = std::find(buffer.begin(), buffer.end(), '\003');
-    if (a != buffer.end()) {
-        return std::string(&buffer[0], a - buffer.begin());
-    }
-    return std::string(&buffer[0], returned_length);
+
+    // Drop contents behind \003, this trick is there to skip certain arguments. (???)
```

Comment:
This feature was introduced in 6c1ddab491b666e125f9d404cdeccf1f84fd81c8 (PR 2289).

It was used to format strings like: `_("%1$s hits %4$s") + "\003<%2$c%3$c>"` - the 2. and 3. parameters that would later be given to `format_string` would otherwise not be used in the format string.

The string `\003` does not appear anywhere in the code (except here), so it can probably be removed.